db/model/Gdoc/GdocFactory.ts

```diff
@@ -1,154 +1,556 @@
+        ...enrichedRow,
+        tags: null,
+    } satisfies OwidGdocBaseInterface
+    if (fetchLinkedTags) {
+        const tags = await knexRaw<DbPlainTag>(
+            knex,
+            `-- sql
+                SELECT tags.*
+                FROM tags
+                JOIN posts_gdocs_x_tags gt ON gt.tagId = tags.id
+                WHERE gt.gdocId = ?`,
+            [gdoc.id]
+        )
+        gdoc.tags = tags
+    }
+    return gdoc
+}
+
+export async function getAndLoadGdocBySlug(
+    knex: KnexReadonlyTransaction,
+    slug: string
+): Promise<GdocPost | GdocDataInsight | GdocHomepage | GdocAuthor> {
+    const base = await getGdocBaseObjectBySlug(knex, slug, true)
+    if (!base) {
+        throw new Error(
+            `No published Google Doc with slug "${slug}" found in the database`
+        )
+    }
+    return loadGdocFromGdocBase(knex, base)
+}
+
+export async function getAndLoadGdocById(
+    knex: KnexReadonlyTransaction,
+    id: string,
+    contentSource?: GdocsContentSource
+): Promise<GdocPost | GdocDataInsight | GdocHomepage | GdocAuthor> {
+    const base = await getGdocBaseObjectById(knex, id, true)
+    if (!base)
+        throw new Error(`No Google Doc with id "${id}" found in the database`)
+    return loadGdocFromGdocBase(knex, base, contentSource)
+}
+
+// From an ID, get a Gdoc object with all its metadata and state loaded, in its correct subclass.
+// If contentSource is Gdocs, use live data from Google, otherwise use the data in the DB.
+export async function loadGdocFromGdocBase(
+    knex: KnexReadonlyTransaction,
+    base: OwidGdocBaseInterface,
+    contentSource?: GdocsContentSource
+): Promise<GdocPost | GdocDataInsight | GdocHomepage | GdocAuthor> {
+    const type = get(base, "content.type") as unknown
+    if (!type)
+        throw new Error(
+            `Database record for Google Doc with id "${base.id}" has no type`
+        )
+    if (!checkIsOwidGdocType(type)) {
+        throw new Error(
+            `Database record for Google Doc with id "${base.id}" has invalid type "${type}"`
+        )
+    }
+
+    const gdoc = match(type)
+        .with(
+            P.union(
+                OwidGdocType.Article,
+                OwidGdocType.LinearTopicPage,
+                OwidGdocType.TopicPage,
+                OwidGdocType.Fragment,
+                OwidGdocType.AboutPage
+            ),
+            () => GdocPost.create(base)
+        )
+        .with(OwidGdocType.DataInsight, () => GdocDataInsight.create(base))
+        .with(OwidGdocType.Homepage, () => GdocHomepage.create(base))
+        .with(OwidGdocType.Author, () => GdocAuthor.create(base))
+        .exhaustive()
+
+    if (contentSource === GdocsContentSource.Gdocs) {
+        // TODO: if we get here via fromJSON then we have already done this - optimize that?
+        await gdoc.fetchAndEnrichGdoc()
+    }
+
+    await gdoc.loadState(knex)
+
+    return gdoc
+}
+
+export async function getAndLoadPublishedDataInsights(
```

Comment:
We have `getPublishedDataInsights` in `db.ts` - is it possible to incorporate that here? Either reuse it in this or use this instead of that?

Reply:
That seems sufficiently different though? Maybe we could improve the naming and eventually we'll maybe want to move the functions around a bit but the db.getPublishedDataInsights fetches only a subset of columns.